### PR TITLE
Add OAuthApiConnection

### PR DESCRIPTION
**Pull request by wetneb**
This is a WIP for #268. I will not have time to work on this soon so I am posting it as is - feel free to take the PR over.

Issues to be discussed:
* I think WDTK should not handle the OAuth login process itself, it should just take the four OAuth tokens (client key, client secret, access key, access secret) and sign its requests with that. The OAuth login process itself should be handled via the application using WDTK using whatever library they like. But ScribeJava is not really designed to be used just to sign requests as far as I can tell, you are forced to configure the OAuth endpoints as far as I can tell (even if they are not relevant for the signing of requests). So maybe ScribeJava isn't a great fit and something lighter should be used (or do you want to include the login logic as well? at least we should make sure users can provide their own tokens.) @lucaswerkmeister, your work on ScribeJava would not be lost - I just think it should be used downstream (for instance in OpenRefine).
* no idea how to test / mock this properly
* how do you log people out? I think it is better to also force a log out from MediaWiki, not just forget their tokens locally (otherwise that makes it hard to switch users: users need to manually go to MediaWiki to log themselves out in addition to logging out from the app).

### CODE REVIEW

**Comment by Tpt** [comment]
> I think WDTK should not handle the OAuth login process itself, it should just take the four OAuth tokens (client key, client secret, access key, access secret) and sign its requests with that.

I would tend to disagree with you here. It would be great to have an interface like the following, that would be easy to use and hide the dependency to ScribeJava. I hope it would cover all use cases and avoid effort duplication on the side of the WikidataToolkit clients.

```
public class OAuthApiConnection extends ApiConnection {
        // The first two parameters are the parameters of com.github.scribejava.apis.MediaWikiApi
	public OAuthApiConnection(String indexUrl, String niceUrlBase, String apiBaseUrl, StringString clientToken, String clientSecret);

        //Returns the authorization URL for this wiki. The secret is generated by the class and kept in its state
        public String getAuthorizationUrl();

        //Gives back the request token to this class that could retrieve the authorization token
        public void setRequestToken(String token);
}
```


> how do you log people out? I think it is better to also force a log out from MediaWiki, not just forget their tokens locally (otherwise that makes it hard to switch users: users need to manually go to MediaWiki to log themselves out in addition to logging out from the app).

Genuine question: is it possible to log out people from MediaWiki with only an OAuth token? If yes, it would be nice to make the logout() method do both indeed. If no, I believe it's fine to just forget their tokens, we could hardly do better.

**Comment by wetneb** [comment]
Yes in any case, of course ScribeJava should not be exposed in the public methods of `OAuthApiConnection`. I am not fundamentally against providing the methods you mention above, that being said I think it would also be important to make sure an `OAuthApiConnection` can be created simply based on the four tokens (client key, client secret, access key, access secret) required to sign the requests. This is required in the following use cases:
* I write a bot and I want to use my own owner-only OAuth client to login (for instance for the sake of having the edits tagged appropriately and avoiding to expose my password there). I just do the OAuth login process once with some tool and just give the tokens to the bot.
* I write an application which already provides OAuth functionality (with Wikibase or another provider), and I do not have full control over the existing login framework
* I just want to be able to serialize and restore later OAuthApiConnection objects (pretty basic requirement I think!)

Also integrating the entire login process in OAuthApiConnection requires to pull in the full ScribeJava (including all the configurations for many third-party providers), whereas sticking to the request signing only requires the inclusion of the core package (or probably an even lighter package).

**Comment by Tpt** [comment]
@wetneb Indeed. These are valid use cases to support. We could go the simplest way: get the tokens from the constructor or a setter and just sign the requests with them, depending on ScribeJava or just on our own simple signing implementation. With this option we let the library users handle all the token retrieval themselves.

**Comment by Tpt** [comment]
@wetneb Indeed. These are valid use cases to support. We could go the simplest way: get the tokens from the constructor or a setter and just sign the requests with them, depending on ScribeJava or just on our own simple signing implementation. With this option we let the library users handle all the token retrieval themselves.

**Comment by Tpt** [comment]
@wetneb Do you want to move forward on it soon? I believe it would be nice to get a new WikidataToolkit release soon with the `UnsupportedValue` system

**Comment by wetneb** [comment]
@Tpt yes the UnsupportedValues should become useful given the introduction of the EntitySchema datatype on Wikidata.

If I can find some time to work more on WDTK soon, I think my priority would be #426 / #412, as this is more likely to benefit OpenRefine immediately.

OAuth support would only be useful for hosted instances of OR, for which a hell of a lot of other issues need to be addressed first in my opinion (such as being able to work on a project concurrently - which requires big architectural changes in the core datamodel of the tool).

**Comment by Tpt** [comment]
@wetneb Ok ! Thank you ! EntitySchema datatype seems to use the DataValue type "string" so, nothing should break for it. If OAuth is not a priority for OpenRefine, let's not block new releases on this PR except if someone want to step up to do it.

**Comment by wetneb** [comment]
Yes at the moment EntitySchema uses "string" but it was suggested that it should rather behave like lexemes / forms / senses, so I suspect it is likely to evolve before being rolled out:
https://www.wikidata.org/wiki/Wikidata:Project_chat#New_data_type_for_EntitySchemas

**Comment by afkbrb** [comment]
@wetneb @Tpt What are the prerequisites to get this branch merged?

**Comment by afkbrb** [comment]
Hi @wetneb , I'm ready to work on this, please assign it to me.

**Comment by afkbrb** [comment]
I think using https://github.com/pakerfeldt/okhttp-signpost is terrific here. 

1. It focus on signing, while scribejava aims to handle all OAuth process.
2. We will use okhttp here, it's helpful for #492 .

But if we want to show how to use the OAuth functionality to be implemented by `OAuthApiConnection`, we will probably need to import scribejava in the examples module.

**Comment by afkbrb** [comment]
@wetneb @Tpt Do we still need to introduce `UsernamePasswordApiConnection` as planned in #268 ?

To do that, we will need to throw an UnsupportedOperationException if the user try to login with `BasicApiConnection`, which will make it not backwards compatible.

**Comment by wetneb** [comment]
I think using okhttp-signpost would be a nice way to minimize the dependencies pulled by this library. Using ScribeJava directly as a dependency to handle the entire OAuth process would also pull in adaptors for tons of OAuth providers: it is probably not super heavy, but it does not feel as clean, in my opinion. But I do not have a strong opinion about this.

I thought that the `BasicApiConnection` that we already have was already meant to be only used for username/password login, but reading #268 again I see Lucas saw it as a common base class for both OAuth and password login. Given the current deprecation annotations that point to `BasicApiConnection` I would expect that class to be something usable directly by the user, not an abstract base class, so if such a base class is needed I would probably reintroduce it under a different name (`ApiConnectionBase` or something similar?).

**Comment by afkbrb** [comment]
So I guess it's better to keep the current `BasicApiConnection`  and not introduce the `UsernamePasswordApiConnection` for backward compatibility and less complexity.

**Comment by wetneb** [comment]
Thanks, fixed!

**Comment by wetneb** [comment]
Hmm, I did not realize that there are merge conflicts, in this PR. Sorry about that @afkbrb, it would have been easier for you to start from scratch instead of salvaging my ancient PR. I can solve these conflicts in the coming days.

**Comment by wetneb** [comment]
I attempted a merge and I hope I got everything right, it was a bit of a can of worms. @afkbrb, could you revriew it? Perhaps use your testing environment to double-check that this still works as expected?

The main conflicts were:
- introduction of token storage in the ApiConnection class
- fix of handling of cookies when response headers have mixed case
- introduction of connect and read timeouts. These are not honoured yet by OAuthApiConnection, it would be good to fix this.

As a follow up PR we should migrate the BasicApiConnection class to use OkHTTP as well.

@Tpt are you happy with the direction this is going?

**Comment by afkbrb** [comment]
Just tested it, still working :)
wikidata-toolkit-oauth-edit

> introduction of connect and read timeouts. These are not honoured yet by OAuthApiConnection, it would be good to fix this.

I'll work on it.

**Comment by afkbrb** [comment]
> As a follow up PR we should migrate the BasicApiConnection class to use OkHTTP as well.

Hi @wetneb , do you mean that we need to migrate `BasicApiConnnection` to use OkHTTP first before merging this branch?

**Comment by wetneb** [comment]
> Hi @wetneb , do you mean that we need to migrate BasicApiConnnection to use OkHTTP first before merging this branch?

No, I think it would be a natural next step, but I think it is simpler to keep this PR small: there are already enough moving pieces here.

**Comment by wetneb** [comment]
@Tpt this is ready for your review! :)

File: wdtk-wikibaseapi/pom.xml [hunk 1]
```diff
@@ -32,6 +32,18 @@
...
+		<dependency>
+			<groupId>com.squareup.okhttp3</groupId>
+			<artifactId>mockwebserver</artifactId>
+			<version>4.2.2</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>se.akerfeldt</groupId>
+			<artifactId>okhttp-signpost</artifactId>
+			<version>1.1.0</version>
```
[the remark this answers is not in the document]
Comment (afkbrb) [reply wdtk-wikibaseapi/pom.xml:44]: Your're right, I'll update it and also other code you mentioned below.

[the remark this answers is not in the document]
Comment (afkbrb) [reply wdtk-wikibaseapi/pom.xml:44]: Hi @Tpt ,I've made the changes in #504 .

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/OAuthApiConnection.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        accessToken = null;
+        accessSecret = null;
+        username = "";
+        loggedIn = false;
+    }
+
+    /**
+     * Checks if the OAuth credentials (i.e., consumer key/secret and access token/secret) are still valid.
+     * <p>
+     * The OAuth credentials can be invalid if the user invoked it.
```
Comment (tfmorris) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/OAuthApiConnection.java:146]: I'm guessing this should says "user revoked them", not "invoked"

**Pull request merged**